### PR TITLE
Qualify the type when negating a pattern to a not pattern

### LINKED ISSUE #72370: Use Pattern Matching suggestion gives faulty code
**Version Used**: C# 12

**Steps to Reproduce**:

1. Paste the following code into Visual Studio:
```cs
public class Program {
    public class Cat {
    }
    public static void Main() {
        Cat Cat = new();
        object Object = Cat;
        
        // Suggestion: IDE0083: Use Pattern Matching
        if (!(Object is Cat)) {
            
        }
    }
}
```

2. If you accept the suggestion, it corrects your code to:
```cs
if (Object is not Cat) {
    
}
```
which gives you a compilation error: `(line 14, col 21): A constant value is expected`

**Diagnostic Id**: IDE0083

**Expected Behavior**:
No suggestion given.

**Actual Behavior**:
Faulty suggestion given.

**Related discussion:**
https://github.com/dotnet/csharplang/discussions/7980

### PR DESCRIPTION
**Pull request by sdelarosbil**
Closes #72370 

When we run into a case like in this issue, we want to qualify the type name such that it resolves the ambiguity and let the simplifier reduce it to the shortest available.

However, simply using the full type expression wasn't enough: it turns out that a not pattern assembled this way requires the member expression version of the node instead of the qualified name one and that the pattern is a constant one instead of a type one. This is because unless it's a predefined type, it's seen as an expression rather than a type name.

With both of these changes, it's enough to resolve the ambiguities.

### CODE REVIEW

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Extensions/SyntaxGeneratorExtensions_Negate.cs [hunk 1]
```diff
@@ -158,6 +158,22 @@ internal static partial class SyntaxGeneratorExtensions
...
+                // object O = C;
+                // if (O is not C)
+                // ...
+                // Then there is an ambiguity between the type C and the local named C.
+                // To address this, we use an expression with the fully qualified type name which the simplifier
+                // will shorten to the shortest possible without ambiguity
+                if (operation is IIsTypeOperation { TypeOperand.SpecialType: SpecialType.None } isTypeOperation &&
+                    syntaxFacts.SupportsNotPattern(semanticModel.SyntaxTree.Options))
+                {
+                    var typeNode = generatorInternal.Type(isTypeOperation.TypeOperand, false);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Extensions/SyntaxGeneratorExtensions_Negate.cs:173]: named arg for `false`

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Extensions/SyntaxGeneratorExtensions_Negate.cs [hunk 2]
```diff
@@ -158,6 +158,22 @@ internal static partial class SyntaxGeneratorExtensions
...
+                    return generatorInternal.IsPatternExpression(leftOperand, operatorToken, generatorInternal.NotPattern(generatorInternal.ConstantPattern(typeNode)));
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Extensions/SyntaxGeneratorExtensions_Negate.cs:174]: should this be a `TypePattern` instead of a `ConstantPattern`?  Or is it actually supposed to be a ConstantPattern from teh perspective of the parser?

Comment (CyrusNajmabadi) [reply src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Extensions/SyntaxGeneratorExtensions_Negate.cs:174]: i don't really see what the `IIsTypeOperation { TypeOperand.SpecialType: SpecialType.None } isTypeOperation` accomplishes.  What if someone has an  `Int32 Int32 = 0` line and does `!(o is Int32)`.  Won't we have the same problem?

Comment (sdelarosbil) [reply src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Extensions/SyntaxGeneratorExtensions_Negate.cs:174]: I originally tried TypePattern and it turns out this is not only what cuased the simplifier issue I mentioned in the issue thread, but I found out later that it actually messes up the expected syntax tree.

It means that yes, constant pattern is what's supposed to be there from the parser's perspective.

Comment (CyrusNajmabadi) [reply src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Extensions/SyntaxGeneratorExtensions_Negate.cs:174]: Ok.  So why not always make a constant pattern.  Or make it a constant pattern if it's a NameSyntax?  The IIs Check seems busted to me since we're making syntax checks, but trying to use semantics to drive it.

Comment (sdelarosbil) [reply src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Extensions/SyntaxGeneratorExtensions_Negate.cs:174]: I tried to make it always a constant pattern and it wouldn't work because you get the same problem where if the right operand is "string", it messes up the syntax tree where it expected a type syntax when parsing

Comment (sdelarosbil) [reply src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Extensions/SyntaxGeneratorExtensions_Negate.cs:174]: > i don't really see what the `IIsTypeOperation { TypeOperand.SpecialType: SpecialType.None } isTypeOperation` accomplishes. What if someone has an `Int32 Int32 = 0` line and does `!(o is Int32)`. Won't we have the same problem?

I checked just now: both of these lines are invalid. It seems you can't do this for these types. If you have an example that compiles, I could check.

Comment (sdelarosbil) [reply src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Extensions/SyntaxGeneratorExtensions_Negate.cs:174]: > Ok. So why not always make a constant pattern. Or make it a constant pattern if it's a NameSyntax? The IIs Check seems busted to me since we're making syntax checks, but trying to use semantics to drive it.

The best I can do is IsSimpleName it seems.

However, even if I do this, I still need the symbol to get the node in MemberAccessForm (the Type method). Is that still fine?

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Extensions/SyntaxGeneratorExtensions_Negate.cs (hunk 1 above)
Comment (sdelarosbil) [reply src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Extensions/SyntaxGeneratorExtensions_Negate.cs]: done.

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Extensions/SyntaxGeneratorExtensions_Negate.cs (hunk 2 above)
Comment (sharwell) [reply src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Extensions/SyntaxGeneratorExtensions_Negate.cs:174]: > should this be a TypePattern instead of a ConstantPattern?

If the tests for the affected code fixes use the new testing library, it will validate that the nodes produced by the code fix are the same as the nodes produced by the compiler during a parsing operation on the same text.

Comment (sdelarosbil) [reply src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Extensions/SyntaxGeneratorExtensions_Negate.cs:174]: @CyrusNajmabadi I am still waiting an answer from you on this review.

Also sidenote, I have more prs than usual pending other than this one (some haven't gotten a review yet)

Comment (CyrusNajmabadi) [reply src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Extensions/SyntaxGeneratorExtensions_Negate.cs:174]: I think you should make the correct syntax based on what you are generating. If you are generating a predefined type syntax, then make the right sort of pattern.  Same for the other type syntaxes.

Comment (CyrusNajmabadi) [reply src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Extensions/SyntaxGeneratorExtensions_Negate.cs:174]: >  I have more prs than usual pending other than this one (some haven't gotten a review yet)

Please ping on prs you'd like eyes on. Thanks.

Comment (sdelarosbil) [reply src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Extensions/SyntaxGeneratorExtensions_Negate.cs:175]: > I think you should make the correct syntax based on what you are generating. If you are generating a predefined type syntax, then make the right sort of pattern. Same for the other type syntaxes.

@CyrusNajmabadi ok I changed the checks to rely more on SyntaxFacts, but there is one confusion that I am not sure is what you wanted so I'll point it out.

I absolutely need to know the TypeOperand of the IIstypeOperation here to generate to appropriate Type syntax.. I didn't see any api that does the same, but on the syntax node instead. In fact, checking this comment tells more or less why the syntax changes depending on this: https://github.com/dotnet/roslyn/blob/991d1f2611877b1fa3666e90fefeea67c30a3599/src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/LanguageServices/SyntaxGeneratorInternalExtensions/SyntaxGeneratorInternal.cs#L76

In other words, I don't know how I can avoid doing a semantic check in any way because I need the actual type symbol to generate the syntax and to do this, as far as I can see, I need to cast the IOperation to IIsTypeOperation. By this codepath, we already confirmed syntaxically that this is an IsTypeExpression so it seems safe to me, but I just want to make sure here because it seems to be the part you had a problem with.

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Extensions/SyntaxGeneratorExtensions_Negate.cs [hunk 3]
```diff
@@ -158,6 +158,23 @@ internal static partial class SyntaxGeneratorExtensions
...
+                // With a not pattern, it is possible to have an ambiguity between a type name or a local identifier.
+                // For example, if a class is named C and we have:
+                // C C = new();
+                // object O = C;
+                // if (O is not C)
+                // ...
+                // Then there is an ambiguity between the type C and the local named C.
+                // To address this, we use an expression with the fully qualified type name which the simplifier
+                // will shorten to the shortest possible without ambiguity
+                if (!syntaxFacts.IsPredefinedType(rightOperand) &&
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Extensions/SyntaxGeneratorExtensions_Negate.cs:170]: i think a better fix is to get the typeNode back, and then determine what sort of pattern needs to be generated based on that typeNode.

Comment (sdelarosbil) [reply src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Extensions/SyntaxGeneratorExtensions_Negate.cs:170]: Sorry, I am confused on how one would do it here: the original expression at best gives me a PredefinedTypeSyntax or an IdentifierNameSyntax. That is the only information I have here to determine which pattern it should be and IsPredefinedType covers this check.

If it's not a predefined type, I have to generate a fully qualified name of the type and wrap it in a constant pattern. The typeNode is only involved there because this is when I know the format it should have.

I think I am missunderstanding because it sounds like you're asking to obtain something I can't get or know if I should get before checking what type rightOperand is.

Comment (CyrusNajmabadi) [reply src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Extensions/SyntaxGeneratorExtensions_Negate.cs:170]: can i make edits to your PR?

Comment (sdelarosbil) [reply src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Extensions/SyntaxGeneratorExtensions_Negate.cs:170]: Sure, go ahead.

**Review by CyrusNajmabadi — APPROVED** [review]

**Comment by CyrusNajmabadi** [comment]
@sdelarosbil Done :)

**Comment by sdelarosbil** [comment]
@CyrusNajmabadi I ran the test that was failing locally before pulling your changes and it passed, but this is definitely a failure caused by the pr.

So it seems your changes broke the test.

**Comment by CyrusNajmabadi** [comment]
Thanks!
